File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        new_list = ListNode(-101)
        curr = new_list
        while list1 or list2:
            if not list1:
                curr.next = list2
                break
            elif not list2:
                curr.next = list1
                break
            elif list1.val <= list2.val:
                curr.next = ListNode(list1.val)
                list1 = list1.next
                curr = curr.next
            else:
                curr.next = ListNode(list2.val)
                list2 = list2.next
                curr = curr.next
        return new_list.next

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head){
        ListNode* prev = NULL;
        ListNode* curr = head;

        while(curr){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        return prev;
    }

    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* curr = head;
        ListNode* lastGroupTail = dummy;
        ListNode* nextGroupFirst;
        int count = 1;
        while(curr){
            ListNode* tail = curr;
            while(count!=k && tail){
                count++;
                tail = tail->next;
            }
            count = 1;
            if(!tail) break;
            nextGroupFirst = tail->next;
            tail->next = nullptr;
            ListNode* reversedHead = reverseList(curr);

            lastGroupTail->next = reversedHead;
            curr->next = nextGroupFirst;
            lastGroupTail = curr;
            curr = nextGroupFirst;
        }
        return dummy->next;
    }
};
